File: BackendAPI/eTFMS/src/main/java/com/eTFMS/eTFMS/model/Fine.java
package com.eTFMS.eTFMS.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "fines")
public class Fine {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(unique = true, nullable = false)
    private String referenceNumber;

    @Column(nullable = false)
    private String categoryIdentifier;

    @Column(nullable = false)
    private String driverId;

    @Column(nullable = false)
    private String officerId;

    @Column(nullable = false)
    private Double amount;

    @Column(nullable = false)
    private LocalDateTime issueDate;

    @Column(nullable = false)
    private LocalDateTime dueDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private FineStatus status;
}

File: MobileApp/etraffic_app/lib/core/models/fine.dart
class Fine {
  final int id;
  final String referenceNumber;
  final int categoryId;
  final String driverId;
  final String officerId;
  final double amount;
  final DateTime issueDate;
  final DateTime dueDate;
  final String status;

  Fine({
    required this.id,
    required this.referenceNumber,
    required this.categoryId,
    required this.driverId,
    required this.officerId,
    required this.amount,
    required this.issueDate,
    required this.dueDate,
    required this.status,
  });

  factory Fine.fromJson(Map<String, dynamic> json) {
    return Fine(
      id: json['id'] as int,
      referenceNumber: json['referenceNumber'] as String,
      categoryId: json['categoryId'] as int,
      driverId: json['driverId'] as String,
      officerId: json['officerId'] as String,
      amount: (json['amount'] as num).toDouble(),
      issueDate: DateTime.parse(json['issueDate'] as String),
      dueDate: DateTime.parse(json['dueDate'] as String),
      status: json['status'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'referenceNumber': referenceNumber,
      'categoryId': categoryId,
      'driverId': driverId,
      'officerId': officerId,
      'amount': amount,
      'issueDate': issueDate.toIso8601String(),
      'dueDate': dueDate.toIso8601String(),
      'status': status,
    };
  }
}
